File: csharp-framework/Core/IPC/Flatmaps/FlatmapList.cs
using System;
using System.Collections;


namespace Metreos.Core.IPC.Flatmaps
{
    /// <summary>A collection which acts like a map which permits duplicate keys</summary>
    /// <remarks>
    /// A list of this type accepts only flatmap-legal data. A key must be a 32-bit integer,
    /// and a value must be one of integer, string, byte[], or a byte[] binary flatmap.   
    /// </remarks>
    /// <remarks>
    /// A list of this type can be converted to and from a binary flatmap. To create a 
    /// FlatmapList from a binary flatmap, pass the flatmap to this list's constructor. 
    /// To create a binary flatmap from a FlatmapList, build a list and invoke its 
    /// ToFlatmap method.  
    /// </remarks>
    [Serializable]
    public class FlatmapList: ICollection, IDictionary, IEnumerable 
    {
        /// <summary>Create an empty FlatmapList</summary>
        public FlatmapList() 
        { 
        }

        /// <summary>Create a read-only FlatmapList from a binary flatmap</summary>
        /// <exception cref="FlatmapException">If flatmap is invalid</exception>
        public FlatmapList(byte[] flatmap)
        {
            Flatmap.FromFlatmap(flatmap, this);
            this.immutable = true;
        }


        /// <summary>Create a possibly modifiable FlatmapList from a binary flatmap</summary>
        /// <exception cref="FlatmapException">If flatmap is invalid</exception>
        public FlatmapList(byte[] flatmap, bool isModifiable)
        {
            Flatmap.FromFlatmap(flatmap, this);
            this.immutable = !isModifiable;
        }


        /// <summary>Convert this list to a binary flatmap</summary>
        /// <exception cref="FlatmapException">If list cannot be converted</exception>
        public byte[] ToFlatmap()
        {
            return this.ToFlatmap(null);
        }


        /// <summary>Convert this list to a binary flatmap</summary>
        /// <param name="headerExtension">Extra header to be embedded if any</param>
        /// <exception cref="FlatmapException">If list cannot be converted</exception>
        public byte[] ToFlatmap(byte[] headerExtension)
        {
            return Flatmap.ToFlatmap(this, headerExtension);
        }


        /// <summary>Precalculate byte length of a flatmap created from current list</summary>
        /// <param name="extraHeaderLength">If a header extension is to be included 
        /// in the calculation, the length of that extra header; otherwise zero</param>
        public int BinaryFlatmapLength(int extraHeaderLength)
        {
            return Flatmap.FlatmapLengthFromList(this, extraHeaderLength);               
        }


        // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
        // List maintenance 
        // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

        /// <summary>Add list entry for specified key and value</summary>
        /// <exception cref="ArgumentException">If object value is not a flatmap type</exception>
        /// <exception cref="InvalidOperationException">If list is read-only</exception>
        public void Add(uint key, object xval)
        {
            if (this.immutable) throw new InvalidOperationException();

            Flatmap.ValueType valueType = Flatmap.ValidateValue(xval);
            if (valueType == Flatmap.ValueType.None)
                throw new ArgumentException("FlatmapList: valueType is None");

            Flatmap.MapEntry entry = new Flatmap.MapEntry(key, valueType, xval);             
            list.Add(entry);
            this.sorted = false;
        }


        /// <summary>Add list entry for specified key and value</summary>
        /// <exception cref="ArgumentException">If object key is not a uint</exception>
        /// <exception cref="ArgumentException">If object value is not a flatmap type</exception>
        /// <exception cref="InvalidOperationException">If list is read-only</exception>
        public void Add(object xkey, object xval)
        {
            if (this.immutable) throw new InvalidOperationException();

            uint key = this.UintKey(xkey);  

            this.Add(key, xval);          
        }


        /// <summary>Remove first list entry for specified key</summary>
        /// <exception cref="ArgumentException">If object key is not a uint</exception>
        /// <exception cref="IndexOutOfRangeException">If entry nonexistent</exception>
        /// <exception cref="InvalidOperationException">If list is read-only</exception>
        public void Remove(object xkey)
        {
            if (this.immutable) throw new InvalidOperationException();

            uint key = this.UintKey(xkey);  
            Flatmap.MapEntry entry = this.Find(key,1);
            if (entry.dataType == Flatmap.ValueType.None)
                throw new IndexOutOfRangeException();

            this.list.Remove(entry);
        }


        /// <summary>Remove list entry at specified index</summary>
        /// <exception cref="IndexOutOfRangeException">If entry nonexistent</exception>
        /// <exception cref="InvalidOperationException">If list is read-only</exception>
        public void RemoveAt(int index)
        {
            if (this.immutable) throw new InvalidOperationException();

            Flatmap.MapEntry entry = this.GetAt(index);
            if (entry.dataType == Flatmap.ValueType.None)
                throw new IndexOutOfRangeException();
          
            this.list.Remove(entry);
        }   


        /// <summary>Remove list entry for specified key and occurence</summary>
        /// <exception cref="IndexOutOfRangeException">If entry nonexistent</exception>
        /// <exception cref="InvalidOperationException">If list is read-only</exception>
        public void Remove(uint key, int occurrence)
        {
            if (this.immutable) throw new InvalidOperationException();

            Flatmap.MapEntry entry = this.Find(key, occurrence);          
            if (entry.dataType == Flatmap.ValueType.None)
                throw new IndexOutOfRangeException();

            this.list.Remove(entry);
        }


        /// <summary>Remove list entry for specified key, value, and occurence</summary>
        /// <exception cref="IndexOutOfRangeException">If entry nonexistent</exception>
        /// <exception cref="InvalidOperationException">If list is read-only</exception>
        public void Remove(uint key, object val, int occurrence)
        {
            if (this.immutable) throw new InvalidOperationException();

            Flatmap.MapEntry entry = this.Find(key, val, occurrence);          
            if (entry.dataType == Flatmap.ValueType.None)
                throw new IndexOutOfRangeException();

            this.list.Remove(entry);
        }


        public void Clear() 
        {   
            this.list.Clear(); 
            this.immutable = this.sorted = false;
        }


        // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
        // List content query 
        // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

        /// <summary>Indicate if list contains specified key</summary>  
        public bool Contains(uint key)
        {
            foreach(Flatmap.MapEntry entry in this.list)             
                if (entry.key == key) return true;             

            return false;
        }


        /// <summary>Indicate if list contains specified value</summary>
        public bool Contains(object xval)
        {
            foreach(Flatmap.MapEntry entry in this.list)             
                 if(entry.dataValue == xval) return true;
              
            return false;           
        }


        /// <summary>Indicate if list contains specified occurrence of specified key</summary>
        public bool Contains(uint key, int occurrence)
        {
            Flatmap.MapEntry entry = this.Find(key, occurrence);          
            return entry.dataType != Flatmap.ValueType.None;
        }


        /// <summary>Indicate if list contains occurrence of specified key and value</summary>
        public bool Contains(uint key, object val, int occurrence)
        {
            Flatmap.MapEntry entry = this.Find(key, val, occurrence);            
            return entry.dataType != Flatmap.ValueType.None;
        }


        /// <summary>Return list entry for specified key and occurence</summary>
        /// <returns>Entry values, or { 0, 0, null } </returns>
        public Flatmap.MapEntry Find(uint key, int occurrence)
        {
            int  occurrencesThisKey = 0;
            int  desiredOccurrence  = occurrence > 0? occurrence: 1;
            this.Sort();

            foreach(Flatmap.MapEntry entry in this.list) 
            {
                if (entry.key < key) continue;
                if (entry.key > key) break;
                if (++occurrencesThisKey < desiredOccurrence) continue; 
                return entry;
            } 

            return new Flatmap.MapEntry();
        }


        /// <summary>Return list entry for specified key, value, and occurence</summary>
        /// <returns>Entry values, or { 0, 0, null } </returns>
        public Flatmap.MapEntry Find(uint key, object val, int occurrence)
        {
            int  occurrencesThisKey = 0;
            int  desiredOccurrence  = occurrence > 0? occurrence: 1;
            this.Sort();

            foreach(Flatmap.MapEntry entry in this.list) 
            {
                if (entry.key < key) continue;
                if (entry.key > key) break;
                if (entry.dataValue != val) continue;
                if (++occurrencesThisKey < desiredOccurrence) continue; 
                return entry;
            } 

            return new Flatmap.MapEntry();
        }


        /// <summary>Return list entry at specified index</summary>
        /// <returns>Entry values, or { 0, 0, null } </returns>
        public Flatmap.MapEntry GetAt(int n)
        {
            this.Sort();
            IEnumerator ix = this.list.GetEnumerator();
            int i = 0;

            while(ix.MoveNext()) 
            {  
                if (i++ != n) continue; 
                Flatmap.MapEntry entry = (Flatmap.MapEntry) ix.Current;
                return entry;
            }

            return new Flatmap.MapEntry();
        }


        public void CopyTo(Array array, int index)
        {
            this.list.CopyTo(array, index);
        }


        public ICollection Keys 
        {   get 
            {   ArrayList keylist = new ArrayList();
                foreach(Flatmap.MapEntry entry in this.list) keylist.Add(entry.key);
                return keylist;
            }
        }


        public ICollection Values 
        {   get 
            {   ArrayList vlist = new ArrayList();
                foreach(Flatmap.MapEntry entry in this.list) vlist.Add(entry.dataValue);
                return vlist;
            }
        }


        IDictionaryEnumerator IDictionary.GetEnumerator()
        {
            return new FlatmapListEnumerator(this.list);
        }


        IEnumerator IEnumerable.GetEnumerator()
        {
            return new FlatmapListEnumerator(this.list);
        }


        // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
        // List sort 
        // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

        public class MapEntryComparer: IComparer  
        {
            int IComparer.Compare(Object x, Object y)  
            {
                Flatmap.MapEntry a = (Flatmap.MapEntry) x;
                Flatmap.MapEntry b = (Flatmap.MapEntry) y;
                return a.key < b.key? -1: a.key > b.key? 1: 0;
            }
        }


        /// <summary>Sorts the list if not currently sorted</summary>
        public void Sort()
        {
            if (!this.sorted) this.list.Sort(new MapEntryComparer());
            this.sorted = true;
        }


        // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
        // Support methods
        // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

        /// <summary>Convert object key to uint</summary>
        /// <exception cref="ArgumentException">If object is not a uint</exception>
        protected uint UintKey(object xkey)
        {
            uint key = 0;
            try  { key = (uint)xkey; }  
            catch{ throw new ArgumentException( "xkey is not uint" ); }
            return key;
        }


        // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
        // Properties
        // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

        protected ArrayList list = new ArrayList();  
        protected bool immutable, sorted;

        public bool IsFixedSize     { get { return false;} }
        public bool IsSynchronized  { get { return false;} }       
        public object SyncRoot      { get { return this; } } 
        public int  Count           { get { return this.list.Count;} }
        public bool IsReadOnly      { get { return this.immutable; } }  
        public object this[int i]   { get { return this.GetAt((int)i); } }
        public object this[object k]
        { get { return null; }  set { throw new InvalidOperationException();} }

        /// <summary>Flatmap header extension</summary>
        /// <remarks>If a custom header extension was serialized with a flatmap,
        /// the header is placed here when the flatmap is deserialized</remarks>
        protected byte[] headerExtension;
        public    byte[] HeaderExtension 
        { get { return headerExtension;  }
          set { headerExtension = value; } 
        }

    } // class FlatmapList



    // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    // FlatmapListEnumerator
    // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

    public class FlatmapListEnumerator: IDictionaryEnumerator
    {
        private int index = -1;
        private ArrayList list;

        internal FlatmapListEnumerator(ArrayList list) { this.list = list; }        

        public bool MoveNext()  { return ++index < list.Count; }
        public void Reset()     { index = -1; }

        public object Current 
        {
            get { return index < 0 || index >= list.Count? null: list[index]; }
        }

        public DictionaryEntry Entry 
        {   get 
            {   Flatmap.MapEntry entry = (Flatmap.MapEntry) this.Current;
                return entry.dataType  == Flatmap.ValueType.None?
                    new DictionaryEntry(null, null):
                    new DictionaryEntry(entry.key, entry.dataValue);            
            }
        }

        public object Key   { get { return Entry.Key;   } }
        public object Value { get { return Entry.Value; } }

    }   //  class FlatmapListEnumerator

}   // namespace

File: java-framework/src/metreos/core/ipc/flatmaps/FlatmapList.java
/* $Id: FlatmapList.java 30151 2007-03-06 21:47:46Z wert $
 *
 * Created by wert on Jan 25, 2005
 *
 * Copyright (c) 2005 Metreos, Inc. All rights reserved.
 */

package metreos.core.ipc.flatmaps;

import java.util.Collections;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;
import java.util.Vector;

/**
 * List of key / value pairs.
 */
public class FlatmapList // : ICollection, IDictionary, IEnumerable 
{
	/**
	 * Constructs an empty FlatmapList.
	 */
	public FlatmapList()
	{
		// nothing to do.
	}

	/**
	 * Constructs a read-only FlatmapList from a binary flatmap.
	 * @param flatmap
	 * @throws FlatmapException
	 */
	public FlatmapList( byte[] flatmap ) throws FlatmapException
	{
		Flatmap.fromFlatmap( flatmap, this );
		immutable = true;
	}

	/**
	 * Constructs a possibly modifiable FlatmapList from a binary flatmap.
	 * @param flatmap
	 * @param isModifiable
	 * @throws FlatmapException
	 */
	public FlatmapList( byte[] flatmap, boolean isModifiable )
			throws FlatmapException
	{
		Flatmap.fromFlatmap( flatmap, this );
		immutable = !isModifiable;
	}

	/**
	 * Converts this list to a binary flatmap with no header extension.
	 * @return the bytes of this flatmap.
	 * @throws FlatmapException
	 */
	public byte[] toFlatmap() throws FlatmapException
	{
		return toFlatmap( null );
	}

	/**
	 * Converts this list to a binary flatmap with the specified header
	 * extension.
	 * @param newHeaderExtension
	 * @return the bytes of this flatmap.
	 * @throws FlatmapException
	 */
	public byte[] toFlatmap( byte[] newHeaderExtension ) throws FlatmapException
	{
		return Flatmap.toFlatmap( this, newHeaderExtension );
	}

	/**
	 * Precalculate byte length of a flatmap created from current list.
	 * @param extraHeaderLength If a header extension is to be included 
	 * in the calculation, the length of that extra header; otherwise zero.
	 * @return length of the resulting flatmap
	 */
	public int binaryFlatmapLength( int extraHeaderLength )
	{
		return Flatmap.flatmapLengthFromList( this, extraHeaderLength );
	}

	// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
	// List maintenance 
	// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

	/**
	 * Adds map entry for specified key and integer value.
	 * @param key
	 * @param nval
	 * @return this flatmap list
	 */
	public FlatmapList add( int key, int nval )
	{
		add( key, new Integer( nval ) );
		return this;
	}
	
	/**
	 * Adds map entry for specified key and long value.
	 * @param key
	 * @param val
	 * @return this flatmap list
	 */
	public FlatmapList add( int key, long val )
	{
		add( key, new Long( val ) );
		return this;
	}
	
	/**
	 * Adds map entry for specified key and double value.
	 * @param key
	 * @param val
	 * @return this flatmap list
	 */
	public FlatmapList add( int key, double val )
	{
		add( key, new Double( val ) );
		return this;
	}
	
	/**
	 * Adds map entry for specified key and object value.
	 * @param key
	 * @param xval
	 * @return this flatmap list
	 */
	public FlatmapList add( int key, Object xval )
	{
		checkMutable();

		int dataType = validateValue( xval );
		if (dataType == VT_NONE)
			throw new IllegalArgumentException( "unrecognized value type" );

		MapEntry entry = new MapEntry( key, dataType, xval );
		list.add( entry );
		sorted = false;
		return this;
	}
	
	/**
	 * Determines flatmap type of supplied value Object.
	 * @param xval
	 * @return flatmap type of supplied value Object. VT_NONE if
	 * the object type is not supported.
	 */
	public static int validateValue( Object xval )
	{
		int type = VT_NONE;

		if (xval instanceof Integer)
			type = VT_INT;
		else if (xval instanceof String)
			type = VT_STRING;
		else if (xval instanceof Long)
			type = VT_LONG;
		else if (xval instanceof Double)
			type = VT_DOUBLE;
		else if (xval instanceof FlatmapList)
			type = VT_FLATMAP;
		else if (xval instanceof byte[])
			type = Flatmap.isFlatmap( (byte[]) xval ) ? VT_FLATMAP : VT_BYTE;

		return type;
	}

	private void checkMutable()
	{
		if (immutable)
			throw new UnsupportedOperationException( "immutable" );
	}

	/**
	 * Adds map entry for specified key and value.
	 * @param xkey
	 * @param xval
	 * @return this flatmap list
	 */
	public FlatmapList add( Object xkey, Object xval )
	{
		add( obj2Int( xkey ), xval );
		return this;
	}

	/**
	 * Removes first map entry for specified key.
	 * @param key
	 */
	public void remove( int key )
	{
		remove( key, 1 );
	}

	/**
	 * Removes first map entry for specified key.
	 * @param xkey
	 */
	public void remove( Object xkey )
	{
		remove( obj2Int( xkey ) );
	}

	/**
	 * Removes map entry at specified index.
	 * @param index
	 */
	public void removeAt( int index )
	{
		checkMutable();

		MapEntry entry = getAt( index );
		if (entry.dataType == VT_NONE)
			throw new IndexOutOfBoundsException();

		list.remove( entry );
		// sorted = false; remove from a sorted list leaves the list sorted
	}

	/**
	 * Removes map entry for specified key and occurrence.
	 * @param key
	 * @param occurrence
	 */
	public void remove( int key, int occurrence )
	{
		checkMutable();

		MapEntry entry = find( key, occurrence );
		if (entry.dataType == VT_NONE)
			throw new IndexOutOfBoundsException();

		list.remove( entry );
		// sorted = false; remove from a sorted list leaves the list sorted
	}

	/**
	 * Removes map entry for specified key, value, and occurrence.
	 * @param key
	 * @param xval
	 * @param occurrence
	 */
	public void remove( int key, Object xval, int occurrence )
	{
		checkMutable();

		MapEntry entry = find( key, xval, occurrence );
		if (entry.dataType == VT_NONE)
			throw new IndexOutOfBoundsException();

		list.remove( entry );
		// sorted = false; remove from a sorted list leaves the list sorted
	}

	/**
	 * Clears the entire list.
	 */
	public void clear()
	{
		checkMutable();
		list.clear();
		// sorted = false; a cleared list is sorted!
	}

	// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
	// List content query 
	// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

	/**
	 * Indicates if list contains specified value.
	 * @param xval
	 * @return true if value exists for any key.
	 */
	public boolean contains( Object xval )
	{
		Iterator<MapEntry> i = list.iterator();
		while (i.hasNext())
		{
			MapEntry entry = i.next();
			if (entry.dataValue.equals( xval ))
				return true;
		}
		return false;
	}

	/**
	 * Indicates if list contains specified key.
	 * @param key
	 * @return true if key exists.
	 */
	public boolean containsKey( int key )
	{
		return containsKey( key, 1 );
	}

	/**
	 * Indicates if list contains specified occurrence of specified key.
	 * @param key
	 * @param occurrence
	 * @return true if key exists.
	 */
	public boolean containsKey( int key, int occurrence )
	{
		return find( key, occurrence ).dataType != VT_NONE;
	}

	/**
	 * Indicates if list contains occurrence of specified key and value.
	 * @param key
	 * @param xval
	 * @param occurrence
	 * @return true if key/value/occurrence exists.
	 */
	public boolean contains( int key, Object xval, int occurrence )
	{
		return find( key, xval, occurrence ).dataType != VT_NONE;
	}

	/**
	 * Returns map entry for specified key and occurrence.
	 * @param key
	 * @param occurrence
	 * @return the entry for the specified key/occurrence, or null.
	 */
	public MapEntry find( int key, int occurrence )
	{
		int occurrencesThisKey = 0;
		int desiredOccurrence = occurrence > 0 ? occurrence : 1;
		sort();

		Iterator<MapEntry> i = list.iterator();
		while (i.hasNext())
		{
			MapEntry entry = i.next();
			if (entry.key < key)
				continue;
			if (entry.key > key)
				break;
			if (++occurrencesThisKey < desiredOccurrence)
				continue;
			return entry;
		}

		return null;
	}

	/**
	 * Returns map entry for specified key, value, and occurrence.
	 * @param key
	 * @param xval
	 * @param occurrence
	 * @return the entry for the specified key/value/occurrence, or null.
	 */
	public MapEntry find( int key, Object xval, int occurrence )
	{
		int occurrencesThisKey = 0;
		int desiredOccurrence = occurrence > 0 ? occurrence : 1;
		sort();

		Iterator<MapEntry> i = list.iterator();
		while (i.hasNext())
		{
			MapEntry entry = i.next();
			if (entry.key < key)
				continue;
			if (entry.key > key)
				break;
			if (!entry.dataValue.equals( xval ))
				continue;
			if (++occurrencesThisKey < desiredOccurrence)
				continue;
			return entry;
		}

		return null;
	}

	/**
	 * Returns map entry at specified index.
	 * @param index
	 * @return the entry for the specified index, or null.
	 */
	public MapEntry getAt( int index )
	{
		sort();
		
		Iterator<MapEntry> i = list.iterator();
		int j = 0;
		while (i.hasNext())
		{
			MapEntry entry = i.next();
			if (j++ != index)
				continue;
			return entry;
		}

		return null;
	}

	@Override
	public String toString()
	{
		sort();
		return list.toString();
	}

	/**
	 * A FlatmapList map entry.
	 */
	public static class MapEntry
	{
		/**
		 * The key of the entry.
		 */
		public final int key;
	
		/**
		 * The data type of the entry.
		 */
		public final int dataType;
	
		/**
		 * The value of the entry.
		 */
		public final Object dataValue;
	
		/**
		 * Constructs a dummy map entry.
		 */
		public MapEntry()
		{
			this( 0, VT_NONE, null );
		}
		
		/**
		 * Constructs a map entry.
		 * 
		 * @param key
		 * 
		 * @param dataType
		 * 
		 * @param dataValue
		 */
		public MapEntry( int key, int dataType, Object dataValue )
		{
			this.key = key;
			this.dataType = dataType;
			this.dataValue = dataValue;
		}

		@Override
		public String toString()
		{
			return ""+key+"("+dataType2String(dataType)+")="+dataValue;
		}

		/**
		 * @return the string value of this map entry.
		 */
		public String stringValue()
		{
			return (String) dataValue;
		}

		/**
		 * @return the integer value of this map entry.
		 */
		public Integer integerValue()
		{
			return (Integer) dataValue;
		}

		/**
		 * @return the flat map list value of this map entry.
		 */
		public FlatmapList flatmapListValue()
		{
			return (FlatmapList) dataValue;
		}
	}

	// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
	// List sort 
	// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

	/**
	 * Comparator for map entries.
	 */
	public static class MapEntryComparer implements Comparator<MapEntry>
	{
		public int compare( MapEntry a, MapEntry b )
		{
			if (a.key < b.key)
				return -1;
			if (a.key > b.key)
				return 1;
			return 0;
		}
	}

	/**
	 * Sorts the list if not currently sorted.
	 */
	void sort()
	{
		if (!sorted)
			Collections.sort( list, new MapEntryComparer() );
		sorted = true;
	}

	// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
	// Support methods
	// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

	private int obj2Int( Object obj )
	{
		return ((Integer) obj).intValue();
	}

	// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
	// Properties
	// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

	private List<MapEntry> list = new Vector<MapEntry>();

	private boolean immutable;
	
	private boolean sorted;
	
	private byte[] headerExtension;

	/**
	 * Returns a name for a given data type.
	 * 
	 * @param dataType a data type.
	 * 
	 * @return the data type's name.
	 */
	public static String dataType2String( int dataType )
	{
		switch (dataType)
		{
			case VT_NONE:
				return "None";
			case VT_STRING:
				return "String";
			case VT_BYTE:
				return "Byte";
			case VT_FLATMAP:
				return "Flatmap";
			case VT_INT:
				return "Int";
			case VT_LONG:
				return "Long";
			case VT_DOUBLE:
				return "Double";
			default:
				return "Unknown";
		}
	}

	/**
	 * Minimum defined data type code.
	 */
	public static final int VT_MIN = 0;

	/**
	 * Data type code for an unknown value type (0).
	 */
	public static final int VT_NONE = 0;

	/**
	 * Data type code for a 32-bit integer (1).
	 */
	public static final int VT_INT = 1;

	/**
	 * Data type code for a byte array (2).
	 */
	public static final int VT_BYTE = 2;

	/**
	 * Data type code for a string (3).
	 */
	public static final int VT_STRING = 3;

	/**
	 * Data type code for a flatmap (4).
	 */
	public static final int VT_FLATMAP = 4;
	
	/**
	 * Data type code for a 64-bit integer (5).
	 */
	public static final int VT_LONG = 5;
	
	/**
	 * Data type code for a 64-bit IEEE floating (6).
	 */
	public static final int VT_DOUBLE = 6;

	/**
	 * Maximum defined data type code.
	 */
	public static final int VT_MAX = 6;
	
	/**
	 * Returns an iterator over all the list entries.
	 * @return an iterator over all the list entries.
	 */
	public Iterator<MapEntry> iterator()
	{
		return list.iterator();
	}

	/**
	 * Counts the number of map entries altogether.
	 * 
	 * @return the number of map entries altogether.
	 */
	public int count()
	{
		return list.size();
	}

	/**
	 * Returns the headerExtension.
	 * 
	 * @return returns the headerExtension.
	 */
	public byte[] getHeaderExtension()
	{
		return headerExtension;
	}

	/**
	 * Sets the headerExtension.
	 * 
	 * @param headerExtension The headerExtension to set.
	 */
	public void setHeaderExtension( byte[] headerExtension )
	{
		this.headerExtension = headerExtension;
	}

	/**
	 * Gets the map entries for the specified key.
	 * 
	 * @param key
	 * 
	 * @return an iterator over the map entries for the
	 * specified key.
	 */
	public Iterator<MapEntry> getValues( int key )
	{
		List<MapEntry> v = new Vector<MapEntry>();
		Iterator<MapEntry> i = list.iterator();
		while (i.hasNext())
		{
			MapEntry me = i.next();
			if (me.key == key)
				v.add( me );
		}
		return v.iterator();
	}

	/**
	 * Gets the first string value of the specified key.
	 * @param key
	 * @return the specified string value or null.
	 */
	public String getString( int key )
	{
		return getString( key, 1 );
	}

	/**
	 * Gets the string value of the specified key/occurrence.
	 * @param key
	 * @param occurrence
	 * @return the specified string value or null.
	 */
	public String getString( int key, int occurrence )
	{
		MapEntry me = find( key, occurrence );
		if (me == null)
			return null;
		return me.stringValue();
	}

	/**
	 * Gets the first int value of the specified key.
	 * @param key
	 * @return the specified int value or null.
	 */
	public Integer getInteger( int key )
	{
		return getInteger( key, 1 );
	}

	/**
	 * Gets the int value of the specified key/occurrence.
	 * @param key
	 * @param occurrence
	 * @return the specified int value or null.
	 */
	public Integer getInteger( int key, int occurrence )
	{
		MapEntry me = find( key, occurrence );
		if (me == null)
			return null;
		return me.integerValue();
	}

	/**
	 * Gets the first int value of the specified key.
	 * @param key
	 * @return the specified int value or null.
	 */
	public FlatmapList getFlatmapList( int key )
	{
		return getFlatmapList( key, 1 );
	}

	/**
	 * Gets the int value of the specified key/occurrence.
	 * @param key
	 * @param occurrence
	 * @return the specified int value or null.
	 */
	public FlatmapList getFlatmapList( int key, int occurrence )
	{
		MapEntry me = find( key, occurrence );
		if (me == null)
			return null;
		return me.flatmapListValue();
	}

	/**
	 * @param key
	 * @param dfltValue
	 * @return the value of the key, or the default value if key has no
	 * value.
	 */
	public int getInt( int key, int dfltValue )
	{
		Integer i = getInteger( key );
		if (i == null)
			return dfltValue;
		return i.intValue();
	}
}
